### ISSUE TITLE
Progress on Variational inference

### ISSUE THREAD

**Issue by PhyloStar** [PhyloStar/graphical-models-reading#1]
Found that [Edward](http://edwardlib.org/tutorials/klqp) has better tools for performing stochastic variational inference. 

The paper by [Ranganath et al. ](https://arxiv.org/pdf/1401.0118.pdf) gives a better algorithm for calculating the gradient. 

But, how to calculate that for Sum-product algorithm graph? I need to study Bishop 8.4, Jordan chapter for better understanding. May be Bethe approximation in chapter 4. The [scribe for Xing's course](https://www.cs.cmu.edu/~epxing/Class/10708-16/note/10708_scribe_lecture12.pdf) gives an idea about how to do Sum-product for a tree. That can help me solve the problem.

**Comment by PhyloStar** [comment]
http://blog.shakirm.com/2015/10/machine-learning-trick-of-the-day-4-reparameterisation-tricks/#381_3

https://blog.evjang.com/2016/11/tutorial-categorical-variational.html

https://tiao.io/post/tutorial-on-variational-autoencoders-with-a-concise-keras-implementation/

https://blog.evjang.com/2016/08/variational-bayes.html

Some links that turned out to be useful.

**Comment by PhyloStar** [comment]
Something more:
https://gabrielhuang.gitbooks.io/machine-learning/content/reparametrization-trick.html

Hopefully, Richard's idea of doing reparameterization might be helpful.

**Comment by PhyloStar** [comment]
- Bethe approximation can be used for calculating sum-product likelihood.
- Is mean field approximation related to Bethe approximation?
- Reparameterization will work for Weibull, Lognormal
- Weibull through uniform random number
- Lognormal through standard normal distribution
- Calculation of weibull, lognormal will work using Tensorflow. Check with Richard.
- Evaluation using simulated trees with branch lengths and tree topologies.

**Comment by PhyloStar** [comment]
https://cs.brown.edu/courses/csci2950-p/lectures/2013-04-11_bethe.pdf

**Comment by PhyloStar** [comment]
KL divergence between two weibull distributions is:

https://arxiv.org/pdf/1310.3713.pdf

**Comment by PhyloStar** [comment]
Reparameterization trick

- https://arxiv.org/pdf/1805.08498.pdf
- https://arxiv.org/pdf/1806.01851.pdf
- https://arxiv.org/pdf/1509.01631.pdf
- https://arxiv.org/pdf/1610.02287.pdf

**Comment by PhyloStar** [comment]
Made more progress with respect to calculating the gradients of the distributions w.r.t the parameters.

**Comment by PhyloStar** [comment]
Tziskas paper is overall versatile paper on Variational EM and EM.

Gelman chapter also has a chapter on variational bayes and EM.

**Comment by PhyloStar** [comment]
Then, Bethe's approximation cab be used for calculating the likelihood of words at internal nodes on a fixed tree. Check Wainright and Jordan...
